src/login/controller.ts

```diff
@@ -46,12 +47,14 @@ class LoginController extends Controller {
     }
 
     if (ctx.request.body.totp) {
-        if (!await userService.validateTotp(user, ctx.request.body.totp)) {
-          log(EventType.totpFailed, ctx.ip(), user.id);
-          return this.redirectToLogin(ctx, '', 'Incorrect TOTP code');
-        }
+      if (!await userService.validateTotp(user, ctx.request.body.totp)) {
+        log(EventType.totpFailed, ctx.ip(), user.id);
+        return this.redirectToLogin(ctx, '', 'Incorrect TOTP code');
+      }
     } else if (await userService.hasTotp(user)) {
-        return this.redirectToLogin(ctx, '', 'TOTP token required');
+      return this.redirectToLogin(ctx, '', 'TOTP token required');
+    } else if (!ctx.request.body.totp && await getSetting('totp') === 'required') {
+      return this.redirectToLogin(ctx, '', 'TOTP code is required');
```

Comment:
Hmm, i don't think this is correct yet. The error message also badly describes what the problem is.

I think a better error message might be:

   The system administrator has made TOTP tokens mandatory, but this user did not have a TOTP configured. Login is disabled.

I'm not sure if the new if statement captures this ...

Reply:
Can you be a bit more specific why this is wrong? I have tested with the TOTP value to be "required", and when there isn't any value inside of the TOTP input, it will redirect to login with the error provided

Reply:
I *think* you can just drop the first half of the if statement.

It just looked a bit suspicious to me because it has this extra condition...